server/src/hardware/protocol/z21/clientkernel.cpp

```diff
@@ -572,4 +802,47 @@ void ClientKernel::keepAliveTimerExpired(const boost::system::error_code& ec)
   startKeepAliveTimer();
 }
 
+void ClientKernel::startInactiveDecoderPurgeTimer()
+{
+  assert(ClientConfig::purgeInactiveDecoderInternal > 0);
+  m_inactiveDecoderPurgeTimer.expires_after(boost::asio::chrono::seconds(ClientConfig::purgeInactiveDecoderInternal));
+  m_inactiveDecoderPurgeTimer.async_wait(std::bind(&ClientKernel::inactiveDecoderPurgeTimerExpired, this, std::placeholders::_1));
+}
+
+void ClientKernel::inactiveDecoderPurgeTimerExpired(const boost::system::error_code& ec)
+{
+  if(ec)
+    return;
+
+  const auto purgeTime = std::chrono::steady_clock::now() - std::chrono::seconds(ClientConfig::purgeInactiveDecoderInternal);
+
+  auto it = m_locoCache.begin();
+  while(it != m_locoCache.end())
+  {
+    if(it->second.lastSetTime < purgeTime)
+    {
+      it = m_locoCache.erase(it);
+    }
+    else
+    {
+      it++;
+    }
+  }
+
+  startInactiveDecoderPurgeTimer();
+}
+
+ClientKernel::LocoCache *ClientKernel::getLocoCache(uint16_t dccAddr)
```

Comment:
This method always returns a valid pointer, so it should return a reference.

When pointers are involved you always need to think about `nullptr`, in this case a reference eliminates this :)

Reply:
We should guard against out of memory if cache gets too big...

Reply:
True, but that is rare, cache won't be bigger that approx. ~10000 loco's (DCC address space) worst case, which still isn't a huge amount of memory.

Traintastic currently doesn't handle OS out of memory, it would be nice to stop all trains and power off in that case, however OS out of memory error are rare these days.